pkg/server/catalog/catalog_test.go

```diff
@@ -0,0 +1,252 @@
+				c := hclConfig
+				// Add fakeUpstreamAuthority to configuration
+				c[upstreamauthority.Type] = map[string]HCLPluginConfig{"testUpstream": {}}
+				return c
+			},
+			verifyRepo: func(t *testing.T, repository *Repository) {
+				// An UpstreamAuthority is expected, and it must call fakeUpstreamAuthority method
+				upstream, ok := repository.Catalog.GetUpstreamAuthority()
+				require.True(t, ok)
+				require.NotNil(t, upstream)
+
+				_, err := upstream.MintX509CA(ctx, &upstreamauthority.MintX509CARequest{})
+				require.Error(t, err, "fakeUpstreamAuthority")
+			},
+		},
+		{
+			name: "contains UpstreamAuthority and UpstreamCA",
+			createHclConfig: func() HCLPluginConfigMap {
+				c := hclConfig
+				// Add UpstreamCA and UpstreamAuthority
+				c[upstreamca.Type] = map[string]HCLPluginConfig{"testUpstream": {}}
+				c[upstreamauthority.Type] = map[string]HCLPluginConfig{"testUpstream": {}}
+				return c
+			},
+			err: "only one UpstreamCA or UpstreamAuthority is allowed. Please remove one of them",
+			verifyRepo: func(t *testing.T, repository *Repository) {
+				assert.Fail(t, "Call to verifyRepo is not expected")
+			},
+		},
+		{
+			name: "contains ported UpstreamCA and UpstreamAuthority",
+			createHclConfig: func() HCLPluginConfigMap {
+				c := hclConfig
+				// Add UpstreamCA and UpstreamAuthority
+				c[upstreamca.Type] = map[string]HCLPluginConfig{"testUpstream": {}}
+				c[upstreamauthority.Type] = map[string]HCLPluginConfig{"testUpstream": {}}
+				return c
+			},
+			// Add testUpstream to ported plugins list
+			ported: map[string]bool{"testUpstream": true},
+			err:    "an UpstreamAuthority already exists in plugin's configuration",
+			verifyRepo: func(t *testing.T, repository *Repository) {
+				assert.Fail(t, "Call to verifyRepo is not expected")
+			},
+		},
+	}
+
+	for _, testCase := range testCases {
+		testCase := testCase
+		t.Run(testCase.name, func(t *testing.T) {
+			portedUpstreamCA = testCase.ported
+			hclConfig := testCase.createHclConfig()
+
+			repo, err := Load(ctx, Config{
+				Log: log,
+				GlobalConfig: catalog.GlobalConfig{
+					TrustDomain: "domain.test",
+				},
+				PluginConfig:     hclConfig,
+				IdentityProvider: identityProvider,
+				AgentStore:       agentStore,
+				MetricsService:   metricsService,
+			})
+
+			if testCase.err != "" {
+				require.EqualError(t, err, testCase.err)
+				return
+			}
+
+			require.NoError(t, err)
+			testCase.verifyRepo(t, repo)
+		})
+	}
+}
+
+// createDefaultConfig create a HclPluginConfigMap with the minimal necessary plugins configuration
+func createDefaultConfig() HCLPluginConfigMap {
+	return HCLPluginConfigMap{
+		datastore.Type:    map[string]HCLPluginConfig{"sql": {}},
+		nodeattestor.Type: map[string]HCLPluginConfig{"join_token": {}},
+		noderesolver.Type: map[string]HCLPluginConfig{"noop": {}},
+		keymanager.Type:   map[string]HCLPluginConfig{"memory": {}},
+	}
+}
+
+type fakeUpstreamAuthorityPlugin struct {
```

Comment:
Since the catalog only uses the `Configure` method on these plugin implementations, the server interface can be embedded to drastically simplify the implementation (no need to define the unused functions).

```suggestion
type fakeUpstreamAuthority struct { upstreamauthority.UpstreamAuthorityServer }

func (fakeUpstreamAuthority) Config(context.Context, *spi.ConfigureRequest) (*spi.ConfigureResponse, error) {
```

Reply:
done